feat: use liquibase session lock instead of transaction lock

filipesilva opened the pull request:
Prevent the migration lock class of errors in non-h2 dbs by using a liquibase session lock instead of a transaction lock.
This type of lock would clear automatically when the connection closes for any reason.

Using this approach there is no need for the metabase release lock command except with h2.
The force migration command still is applicable though, as it does more things than just release the lock.

See https://github.com/blagerweij/liquibase-sessionlock

Fix https://github.com/metabase/metabase/issues/50212

Other related issues and PRs: https://github.com/metabase/metabase/issues/50364, https://github.com/metabase/metabase/issues/45298, https://github.com/metabase/metabase/issues/45125, https://github.com/metabase/metabase/issues/44577, https://github.com/metabase/metabase/issues/34486, https://github.com/metabase/metabase/issues/37872, https://github.com/metabase/metabase/issues/37866, https://github.com/metabase/metabase/issues/34551, https://github.com/metabase/metabase/issues/10651

filipesilva committed:
feat: use liquibase session lock instead of transaction lock

Prevent the migration lock class of errors in non-h2 dbs by using a liquibase session lock instead of a transaction lock.
This type of lock would clear automatically when the connection closes for any reason.

Using this approach there is no need for the metabase release lock command except with h2.
The force migration command still is applicable though, as it does more things than just release the lock.

See https://github.com/blagerweij/liquibase-sessionlock

Fix #50212

nvoxland commented on test/metabase/db/liquibase_test.clj:
```diff
@@ -53,14 +53,17 @@
   Ids are orderer in the order it's defined in migration file."
   [file-path db-type]
-  (let [content (u.yaml/from-file (io/resource file-path))]
+  (let [content (u.yaml/from-file (io/resource file-path))
```
Since this no longer applies, we should be able to just delete the changeset from 000_legacy_migrations.yaml rather than add the programatic removal?

filipesilva replied:
Done

nvoxland commented on test/metabase/db/force_migration_test.clj:
```diff
@@ -28,8 +28,8 @@
-;; This test tests 4 things:
-;; - It forces release the locks if there is one
+;; This test tests 4 (on h2) or 3 (on others) things:
+;; - It forces release the locks if there is one (only on h2, others use a session lock that can't be forced)
```
Since only h2 can release locks now, it seems like the `force-release-locks!` function should throw a user-helpful error like "Database X does not support forcing migration locks. The lock is always automatically removed when the process performing the migration completes" for any non-h2 databases?

filipesilva replied:
Done

filipesilva committed:
fix: mark migration 109 as none db

From Nathan Voxland:

From a liquibase-standpoint, deleting the changesets are just fine. As long as you don't accidentally reuse the id+author+path combo in the future. Databases that already ran it will not care that it's missing and new databases will just never see it to know. So it's more of an our-standards questions.

Alternately, you can set dbms="none" to keep it in the file but not have it actually apply anywhere. That doesn't change the checksum

filipesilva committed:
fix: error when force-releasing non-h2 locks

qnkhuat commented on resources/migrations/000_legacy_migrations.yaml:
this file is no longer used for new installs, so if you really need to change this, you'll have to modify `resources/migrations/initialization/metabase_mysql.sql`

filipesilva replied:
This particular migration does not seem to be in that file, but it is used in tests.

crisptrutski commented on resources/migrations/000_legacy_migrations.yaml:
```diff
@@ -5829,16 +5829,14 @@ databaseChangeLog:
   - changeSet:
         id: 109
         author: camsaul
-        comment: Added 0.34.2
```
Since we're changing the checksum we should probably add `validCheckSum: ANY`, even if real world instances are ignoring this file now.

filipesilva replied:
This was a tip from @nvoxland actually:

> Alternately, you can set dbms="none" to keep it in the file but not have it actually apply anywhere. That doesn't change the checksum

(I left that exchange in the commit message)

crisptrutski replied:
TIL! I thought Liquibase was far more aggressive with the checkum

crisptrutski commented on src/metabase/db/liquibase.clj:
```diff
@@ -204,7 +205,10 @@
   [^Liquibase liquibase]
-  (.forceReleaseLocks liquibase))
+  (if-not (= (mdb.connection/db-type) :h2)
+    (log/warnf "Database %s does not support forcing migration locks. The lock is always automatically removed when the process performing the migration completes or fails."
```
Why not make this a failure?

filipesilva replied:
Thought about this a bit more, now I'm pretty sure that it should throw, and instead `release-lock-if-needed!` shouldn't even try if the conn is not h2.

crisptrutski commented on test/metabase/db/force_migration_test.clj:
```diff
@@ -38,13 +38,12 @@
     (mt/with-temp-empty-app-db [conn driver/*driver*]
       (with-redefs [liquibase/changelog-file "force-migration.yaml"]
         (let [data-source  (mdb/data-source)
-              database     (->> (if (instance? java.sql.Connection conn)
```
Do we understand why this overloading was necessary before, to be sure we can remove it?

Just having a Chesterton's fence feeling, as unlikely as it is.

crisptrutski commented on test/metabase/db/liquibase_test.clj:
```diff
@@ -110,7 +110,7 @@
 
 (deftest release-all-locks-if-needed!-test
-  (mt/test-drivers #{:h2 :mysql :postgres}
+  (mt/test-drivers #{:h2}
```
Can we add a test for the other databases using the session lock?

filipesilva replied:
I don't think such a test would mean anything. You can't release the locks anymore from another process, you have to wait or kill the lock-holding process.

crisptrutski replied:
I was thinking something like the following:

1. Start a thread which opens a connection with the lock
2. Have this thread then blocks on a latch.
2. Release the latch from the main thread, to let the thread exit and close its connection.
3. Confirm that you can connect.

A bit fiddly, since you'd have to bypass our connection pool, but without any tests we just have action of a distance when a dependency is added, and wouldn't catch a regression, even if the dependency was removed.

crisptrutski reviewed (approved):
This makes me a little bit nervous, as if my understanding is correct a very expensive migration would continue running even if the initiating process is terminated. What I like about manually resetting is that an operator is able to first check for database activity.

In practice I suppose we were not taking a case-by-case approach in Cloud, so this is better than blind automation?

filipesilva commented:
> if my understanding is correct a very expensive migration would continue running even if the initiating process is terminated.

@crisptrutski this is not correct: the session lock is always released if the initiating process is terminated. That is the case that https://github.com/blagerweij/liquibase-sessionlock covers.

crisptrutski commented:
Yes exactly, my concern is that the lock is released despite, say, an expensive alter tables command still continuing to run for some time.

imrkd commented:
> This makes me a little bit nervous, as if my understanding is correct a very expensive migration would continue running even if the initiating process is terminated.

@crisptrutski I don't think this changes anything and also not what our error messages are advising or what our self hosting customers are doing. (a) it's fixable db-server-side - you can kill off queries on client disconnect via e.g `client_connection_check_interval` as we do in cloud (b) in practice this will almost definitely result in a deadlock.

filipesilva committed:
test: add session lock test

filipesilva commented on test/metabase/db/liquibase_test.clj:
```diff
@@ -128,3 +132,24 @@
             (testing "The lock was released before the migration finished"
               (is (not @locked?)))))))))
+
+(deftest auto-release-session-lock-test
```
@crisptrutski this does it. If you run this test without the dep it will fail.

crisptrutski replied:
Looks great!
